azurerm_nginx_deployment - support for the auto_scale_profile property

puneetsarna opened the pull request:
Acceptance test:

```shell
terraform-provider-azurerm $ make acctests SERVICE='nginx' TESTARGS='-run=TestAccNginxDeployment_autoscaling' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/nginx -run=TestAccNginxDeployment_autoscaling -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccNginxDeployment_autoscaling
=== PAUSE TestAccNginxDeployment_autoscaling
=== CONT  TestAccNginxDeployment_autoscaling
--- PASS: TestAccNginxDeployment_autoscaling (490.74s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/nginx	493.336s
```

puneetsarna commented:
This PR stays as a draft which is pending https://github.com/hashicorp/terraform-provider-azurerm/pull/24868

puneetsarna commented:
TODOs:
- user facing docs.
- Update tests to account for capacity/autoscale ExactlyOneOf.
- Run autoscaling tests.

stephybun commented on website/docs/r/nginx_deployment.html.markdown:
```diff
@@ -154,6 +158,16 @@ A `network_interface` block supports the following:
 
+---
+
+An `auto_scale_profile` block exports the following:
+
+* `name` - (Required) Specify the name of the autoscaling profile.
+
+* `min` - (Required) Specify the minimum number of NGINX capacity units for this NGINX Deployment.
+
+* `max` - (Required) Specify the maximum number of NGINX capacity units for this NGINX Deployment.
```
In the resource documentation for blocks with arguments we use the phrase `supports the following` instead of `exports the following`
```suggestion
An `auto_scale_profile` block supports the following:

* `name` - (Required) Specify the name of the autoscaling profile.

* `min` - (Required) Specify the minimum number of NGINX capacity units for this NGINX Deployment.

* `max` - (Required) Specify the maximum number of NGINX capacity units for this NGINX Deployment.
```

puneetsarna replied:
done 👍

puneetsarna replied:
Going to resolve this as the suggestion was addressed.

stephybun reviewed (changes requested):
Thanks for this PR @puneetsarna. I think there is a breaking change in the way the existing `capacity` property has been modified to support the new `auto_scale_profile` block. I also have some suggestions around property naming and the documentation. If you could take a look and address those comments left in-line then we can take another look through this.

puneetsarna commented on internal/services/nginx/nginx_deployment_resource.go:
```diff
@@ -101,10 +108,37 @@ func (m DeploymentResource) Arguments() map[string]*pluginsdk.Schema {
 			Type:         pluginsdk.TypeInt,
 			Optional:     true,
-			Default:      20,
+			ExactlyOneOf: []string{"auto_scale_profile", "capacity"},
```
Hi @stephybun Thanks for looking into this.

The default capacity that was added here https://github.com/hashicorp/terraform-provider-azurerm/pull/24033 has been causing some issues now that the API supports autoscaling. To be specific, if a user with the `capacity` field set, moves over to use `auto_scale_profile`, applies the terraform change, and then reruns plan, they see an unusual diff where capacity is being changed from 0 -> 20. This is undesirable behavior and an application would succeed and drop autoscaling (and the plan/apply loop continues).

- Is there a way to handle such a behavior in Terraform?
- I understand it's a breaking change, and wanted to know the policy around it as well. You are saying we can only do this on TF version 4.0.0? Given the error handling will call out that the user needs to set the `one of` fields, it should guide them to do so and their deployments would continue to work after that?

stephybun replied:
> Is there a way to handle such a behavior in Terraform?

In cases where setting a new property on a resource changes the value of an existing property with a default, such as this one we advise users to apply `ignore_changes` on the existing property affected when configuring/setting the new property. We usually add this as a note under the property in the documentation.

> I understand it's a breaking change, and wanted to know the policy around it as well. You are saying we can only do this on TF version 4.0.0?

We can only remove the default for `capacity` and set `ExactlyOneOf` on `capacity` and the `auto_scale_profile` block in a major version, so in 4.0. What we usually do is to wrap behaviours that are breaking and should only be applied in the next major release behind the feature flag `features.FourPointOh()`, you can find examples on how to do this by searching for the flag in the code base. It can be a bit tedious to do for schema changes so if you have issues I'm happy to push that change to your branch for you, just let me know.

> Given the error handling will call out that the user needs to set the one of fields, it should guide them to do so and their deployments would continue to work after that?

There are often cases where the Azure API doesn't play nice with the way Terraform works and we have to go down the route of error'ing a users config or allowing a diff in order to guide users to make appropriate changes to fix their deployments. However I don't think that applies here, we can avoid breaking users that haven't set the `capacity` property by wrapping the removal of the default and applying `ExactlyOneOf` behind the feature flag. This will let us support the new `auto_scale_profile` and not force users to make a configuration change if they don't want to configure `auto_scale_profile`.

puneetsarna replied:
Thanks a lot for these details @stephybun 

I chatted with some members of our team internally and we like the idea of going with `ignore_changes` for the time being. I have updated the PR to remove the breaking change. Can you please take a look? Thanks!!

puneetsarna replied:
Going to resolve this given we are not doing a breaking change.

puneetsarna commented on internal/services/nginx/nginx_deployment_resource.go:
```diff
@@ -101,10 +108,37 @@ func (m DeploymentResource) Arguments() map[string]*pluginsdk.Schema {
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+
+					"min": {
+						Type:         pluginsdk.TypeInt,
+						Required:     true,
+						ValidateFunc: validation.IntPositive,
+					},
+
+					"max": {
+						Type:         pluginsdk.TypeInt,
+						Required:     true,
+						ValidateFunc: validation.IntPositive,
+					},
```
I think I like that 👍 I agree that when I wrote this, `min` and `max` sounded confusing as it does not really tell the user the property on which the `min` and `max` applies.

I like your suggestion around `min_capacity` and `max_capacity` and will update this PR with the change.

puneetsarna replied:
Done 👍

puneetsarna commented on internal/services/nginx/nginx_deployment_resource_test.go:
```diff
@@ -281,3 +320,20 @@ resource "azurerm_subnet" "test" {
 `, data.RandomInteger, data.Locations.Primary)
 }
+
+func TestAccNginxDeployment_autoscaling(t *testing.T) {
```
done 👍

puneetsarna commented on internal/services/nginx/nginx_deployment_resource_test.go:
```diff
@@ -131,6 +131,45 @@ resource "azurerm_nginx_deployment" "test" {
+
+
+
+
```
done.

stephybun commented on internal/services/nginx/nginx_deployment_resource_test.go:
```diff
@@ -131,6 +131,94 @@ resource "azurerm_nginx_deployment" "test" {
 
+func (a DeploymentResource) basicAutoscaling(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+
+
+
```
There's some erroneous spacing here 
```suggestion
```

stephybun commented on internal/services/nginx/nginx_deployment_resource_test.go:
```diff
@@ -131,6 +131,94 @@ resource "azurerm_nginx_deployment" "test" {
+
+func (a DeploymentResource) basicAutoscaling_update(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+
+
+
```
Same here 
```suggestion
```

puneetsarna replied:
done 👍  Going to resolve this.

stephybun commented on internal/services/nginx/nginx_deployment_resource.go:
```diff
@@ -99,10 +106,38 @@ func (m DeploymentResource) Arguments() map[string]*pluginsdk.Schema {
-			ValidateFunc: validation.IntPositive,
+			Type:          pluginsdk.TypeInt,
+			Optional:      true,
+			ConflictsWith: []string{"auto_scale_profile"},
```
If the desired behaviour is `ExactlyOneOf` then we usually wrap this around the `features.FourPointOh` flag or at the very least leave a `TODO 4.0` comment so that maintainers know a breaking change needs to be introduced here for the next major release 
```suggestion
			// TODO 4.0 replace this with ExactlyOneOf
			ConflictsWith: []string{"auto_scale_profile"},
```

puneetsarna replied:
👍 Thanks for sharing this. After an internal chat within our team, I think we'd like to leave the breaking change off entirely for now as we are also evaluating other changes as well which might impact some behavior in Terraform. 

Given it's unclear from our end around what those changes might be and the associated timelines, this comment might be confusing and I going to leave it off for the time being.

stephybun reviewed (commented):
Thanks @puneetsarna, a few more minor comments around the spacing in the tests and on leaving a comment on the desired behaviour for the next major release. Once those are addressed this should be good to go!

puneetsarna commented:
Hi @stephybun Thanks for the helping out with the review and for sharing the details around breaking changes. I have addressed all the feedback you shared. Please let me know if there is anything else 👍

stephybun reviewed (approved):
Thanks @puneetsarna LGTM 🍰
